File: mobile/android/app/src/main/java/com/volttracker/obdpoc/data/TelemetrySampleRecord.java
package com.volttracker.obdpoc.data;

import org.json.JSONException;
import org.json.JSONObject;

public final class TelemetrySampleRecord {
    public final long id;
    public final long sessionId;
    public final long capturedAtMs;
    public final String source;
    public final String vehicleState;
    public final int speedKph;
    public final int rpm;
    public final int coolantC;
    public final int loadPct;
    public final int throttlePct;
    public final double voltage;
    public final double soc;
    public final double batteryTemp;
    public final double powerKw;
    public final int sampleNumber;
    public final long sessionMs;
    public final String raw;
    public final String json;

    TelemetrySampleRecord(
            long id,
            long sessionId,
            long capturedAtMs,
            String source,
            String vehicleState,
            int speedKph,
            int rpm,
            int coolantC,
            int loadPct,
            int throttlePct,
            double voltage,
            double soc,
            double batteryTemp,
            double powerKw,
            int sampleNumber,
            long sessionMs,
            String raw,
            String json) {
        this.id = id;
        this.sessionId = sessionId;
        this.capturedAtMs = capturedAtMs;
        this.source = nonNull(source);
        this.vehicleState = nonNull(vehicleState);
        this.speedKph = speedKph;
        this.rpm = rpm;
        this.coolantC = coolantC;
        this.loadPct = loadPct;
        this.throttlePct = throttlePct;
        this.voltage = voltage;
        this.soc = soc;
        this.batteryTemp = batteryTemp;
        this.powerKw = powerKw;
        this.sampleNumber = sampleNumber;
        this.sessionMs = sessionMs;
        this.raw = nonNull(raw);
        this.json = nonNull(json);
    }

    public JSONObject toJson() {
        try {
            return new JSONObject(json);
        } catch (JSONException ex) {
            return new JSONObject();
        }
    }

    private static String nonNull(String value) {
        return value == null ? "" : value;
    }
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/data/TelemetrySampleRecord.kt
package com.volttracker.obdpoc.data

import org.json.JSONException
import org.json.JSONObject

class TelemetrySampleRecord(
    @JvmField val id: Long,
    @JvmField val sessionId: Long,
    @JvmField val capturedAtMs: Long,
    source: String?,
    vehicleState: String?,
    @JvmField val speedKph: Int,
    @JvmField val rpm: Int,
    @JvmField val coolantC: Int,
    @JvmField val loadPct: Int,
    @JvmField val throttlePct: Int,
    @JvmField val voltage: Double,
    @JvmField val soc: Double,
    @JvmField val batteryTemp: Double,
    @JvmField val powerKw: Double,
    @JvmField val sampleNumber: Int,
    @JvmField val sessionMs: Long,
    raw: String?,
    json: String?,
) {
    @JvmField val source: String = nonNull(source)

    @JvmField val vehicleState: String = nonNull(vehicleState)

    @JvmField val raw: String = nonNull(raw)

    @JvmField val json: String = nonNull(json)

    fun toJson(): JSONObject =
        try {
            JSONObject(json)
        } catch (ex: JSONException) {
            JSONObject()
        }
}

private fun nonNull(value: String?): String = value ?: ""
